fcrepo-persistence-ocfl/src/test/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSessionTest.java

```diff
@@ -0,0 +1,654 @@
+        // Change the same file again
+        session.write(FILE1_SUBPATH, fileStream(FILE_CONTENT2));
+        final String versionId = session.commit(NEW_VERSION);
+
+        assertEquals("v1", versionId);
+        assertNoMutableHead(OBJ_ID);
+        assertFileInHeadVersion(OBJ_ID, FILE1_SUBPATH, FILE_CONTENT2);
+    }
+
+    @Test
+    public void write_ReplaceFile_NewVersion_ExistingObject() throws Exception {
+        final var preStagingPath = tempFolder.newFolder("prestage").toPath();
+        Files.writeString(preStagingPath.resolve(FILE1_SUBPATH), FILE_CONTENT1);
+        ocflRepository.putObject(ObjectVersionId.head(OBJ_ID), preStagingPath, new CommitInfo());
+
+        session.write(FILE1_SUBPATH, fileStream(FILE_CONTENT2));
+        final String versionId = session.commit(NEW_VERSION);
+
+        assertEquals("v2", versionId);
+        assertNoMutableHead(OBJ_ID);
+        assertFileInHeadVersion(OBJ_ID, FILE1_SUBPATH, FILE_CONTENT2);
+        assertFileInVersion(OBJ_ID, "v1", FILE1_SUBPATH, FILE_CONTENT1);
+    }
+
+    @Test
+    public void write_ReplaceFileInMHead_ExistingObject() throws Exception {
+        final var preStagingPath = tempFolder.newFolder("prestage").toPath();
+        Files.writeString(preStagingPath.resolve(FILE1_SUBPATH), FILE_CONTENT1);
+        ocflRepository.stageChanges(ObjectVersionId.head(OBJ_ID), new CommitInfo(), updater -> {
+            updater.addPath(preStagingPath, "", MOVE_SOURCE);
+        });
+
+        session.write(FILE1_SUBPATH, fileStream(FILE_CONTENT2));
+        final String versionId = session.commit(MUTABLE_HEAD);
+
+        assertEquals("Multiple commits to head should stay on version", "v2", versionId);
+        assertMutableHeadPopulated(OBJ_ID);
+        assertFileInHeadVersion(OBJ_ID, FILE1_SUBPATH, FILE_CONTENT2);
+    }
+
+    @Test
+    public void writeToMHeadThenNewVersion() throws Exception {
+        // Write first file to mutable head
+        session.write(FILE1_SUBPATH, fileStream(FILE_CONTENT1));
+        session.commit(MUTABLE_HEAD);
+
+        assertMutableHeadPopulated(OBJ_ID);
+
+        stagingPath = tempFolder.newFolder("obj1-staging").toPath();
+        session = new DefaultOCFLObjectSession(OBJ_ID, stagingPath, ocflRepository);
+
+        // Commit changes to new version in second commit
+        session.write(FILE2_SUBPATH, fileStream(FILE_CONTENT2));
+        final String versionId = session.commit(NEW_VERSION);
+
+        assertEquals("v2", versionId);
+        assertNoMutableHead(OBJ_ID);
+        assertFileInHeadVersion(OBJ_ID, FILE1_SUBPATH, FILE_CONTENT1);
+        assertFileInHeadVersion(OBJ_ID, FILE2_SUBPATH, FILE_CONTENT2);
+        // Initial version is empty since created with mutable head
+        assertFileNotInVersion(OBJ_ID, "v1", FILE1_SUBPATH);
+        assertFileNotInVersion(OBJ_ID, "v1", FILE2_SUBPATH);
+    }
+
+    @Test(expected = PersistentSessionClosedException.class)
+    public void write_CommittedSession() throws Exception {
```

Comment:
Just curious about the underscores in the method names...

Reply:
I'm assuming those are just for clarity?

Reply:
I can switch away from this naming scheme if it'd be preferred for consistency. Its to break up the method being tested from state and expectations. There don't seem to be particularly well defined best practices for naming unit tests in java and they can be pretty hard to read. I felt like this was more readable when I came across it while ago while looking for best practices.

Reply:
I think if it improves readability we should keep it.  As Emerson said "A foolish consistency is the hobgoblin of little minds."